chain/state/statetree.go

```diff
@@ -26,16 +26,71 @@ type StateTree struct {
 	root  *hamt.Node
 	Store cbor.IpldStore
 
-	// Maps ID addresses to actors.
-	actorcache map[address.Address]*types.Actor
-	snapshots  []cid.Cid
+	// Maps ID addresses to actors, layered by snapshots
+	snaps *stateSnaps
+}
+
+type stateSnaps struct {
+	layers []map[address.Address]streeOp
+}
+
+type streeOp struct {
+	Act    types.Actor
+	Delete bool
+}
+
+func newStateSnaps() *stateSnaps {
+	return &stateSnaps{
+		layers: []map[address.Address]streeOp{make(map[address.Address]streeOp)},
+	}
+}
+
+func (ss *stateSnaps) addLayer() {
+	ss.layers = append(ss.layers, make(map[address.Address]streeOp))
+}
+
+func (ss *stateSnaps) dropLayer() {
+	ss.layers = ss.layers[:len(ss.layers)-1]
+}
+
+func (ss *stateSnaps) mergeLastLayer() {
+	last := ss.layers[len(ss.layers)-1]
+	nextLast := ss.layers[len(ss.layers)-2]
+
+	for k, v := range last {
+		nextLast[k] = v
+	}
+
+	ss.dropLayer()
+}
+
+func (ss *stateSnaps) getActor(addr address.Address) (*types.Actor, error) {
+	for i := len(ss.layers) - 1; i >= 0; i-- {
+		act, ok := ss.layers[i][addr]
+		if ok {
+			if act.Delete {
+				return nil, types.ErrActorNotFound
+			}
+
+			return &act.Act, nil
+		}
+	}
+	return nil, nil
+}
+
+func (ss *stateSnaps) setActor(addr address.Address, act *types.Actor) {
+	ss.layers[len(ss.layers)-1][addr] = streeOp{Act: *act}
```

Comment:
Technically this can override actor that is marked for deletion, but I don't think this can happen in practice.

And setting a deleted actor should probably 'bring it back' anyways